File: ios/GabpaWang/Data/Pushup/PushUpCounter.swift
import Foundation

enum PushUpState {
    case idle, warming, up, down
}

/// 어깨 Y 좌표 시계열로 푸시업 횟수를 세는 순수 알고리즘.
/// `PushUpCounter.kt`를 1:1 포팅 — 모든 상수와 절차 동일.
final class PushUpCounter {
    private(set) var count: Int = 0
    private(set) var state: PushUpState = .idle
    private(set) var yLow: Float = 0.5
    private(set) var yHigh: Float = 0.5

    private var smoothBuf: [Float] = []
    private var window: [Float] = []
    private var lastSmoothed: Float = 0
    private var stateFrames = 0

    private let SMOOTH_WINDOW = 5
    private let WINDOW_FRAMES = 120     // 4s @ 30fps
    private let MIN_INIT_FRAMES = 45    // ~1.5s warmup
    private let DOWN_FRACTION: Float = 0.45
    private let UP_FRACTION: Float = 0.20
    private let MIN_BAND: Float = 0.06
    private let PERCENTILE_LOW: Float = 0.15
    private let PERCENTILE_HIGH: Float = 0.85
    private let MIN_UP_FRAMES = 9
    private let MIN_DOWN_FRAMES = 11

    func start() {
        state = .warming
        count = 0
        smoothBuf.removeAll(keepingCapacity: true)
        window.removeAll(keepingCapacity: true)
        lastSmoothed = 0
        stateFrames = 0
        yLow = 0.5
        yHigh = 0.5
    }

    func reset() {
        state = .idle
        count = 0
        smoothBuf.removeAll(keepingCapacity: true)
        window.removeAll(keepingCapacity: true)
        lastSmoothed = 0
        stateFrames = 0
        yLow = 0.5
        yHigh = 0.5
    }

    /// 새 어깨 Y 입력. true 반환 시 카운트가 1 증가했다는 뜻.
    @discardableResult
    func process(shoulderY: Float) -> Bool {
        guard state != .idle else { return false }

        let s = smooth(shoulderY)
        lastSmoothed = s
        if window.count >= WINDOW_FRAMES { window.removeFirst() }
        window.append(s)

        if window.count < MIN_INIT_FRAMES {
            stateFrames += 1
            return false
        }

        if state == .warming {
            state = .up
            stateFrames = 0
        }

        let sorted = window.sorted()
        let n = sorted.count
        let yl = sorted[Int(Float(n) * PERCENTILE_LOW)]
        let yh = sorted[Int(Float(n) * PERCENTILE_HIGH)]
        let band = yh - yl

        yLow = yl
        yHigh = yh

        guard band >= MIN_BAND else {
            stateFrames += 1
            return false
        }

        let downThr = yl + DOWN_FRACTION * band
        let upThr   = yl + UP_FRACTION * band

        switch state {
        case .up:
            stateFrames += 1
            if s > downThr && stateFrames >= MIN_UP_FRAMES {
                state = .down
                stateFrames = 0
            }
            return false

        case .down:
            stateFrames += 1
            if stateFrames >= MIN_DOWN_FRAMES && s < upThr {
                state = .up
                count += 1
                stateFrames = 0
                return true
            }
            return false

        default:
            return false
        }
    }

    func currentDisplacement() -> Float { lastSmoothed - yLow }

    func downThresholdValue() -> Float {
        let band = yHigh - yLow
        return yLow + DOWN_FRACTION * band
    }

    func upThresholdValue() -> Float {
        let band = yHigh - yLow
        return yLow + UP_FRACTION * band
    }

    func warmupProgress() -> Float {
        state == .warming ? Float(window.count) / Float(MIN_INIT_FRAMES) : 1
    }

    private func smooth(_ y: Float) -> Float {
        if smoothBuf.count >= SMOOTH_WINDOW { smoothBuf.removeFirst() }
        smoothBuf.append(y)
        let sum = smoothBuf.reduce(0, +)
        return sum / Float(smoothBuf.count)
    }
}

File: app/src/main/java/com/gabpawang/app/PushUpCounter.kt
package com.gabpawang.app

import java.util.ArrayDeque

enum class PushUpState { IDLE, WARMING, UP, DOWN }

class PushUpCounter {
    var count = 0
        private set
    var state = PushUpState.IDLE
        private set
    var yLow = 0.5f
        private set
    var yHigh = 0.5f
        private set

    private val smoothBuf = ArrayDeque<Float>()
    private val window = ArrayDeque<Float>()
    private var lastSmoothed = 0f
    private var stateFrames = 0

    companion object {
        private const val SMOOTH_WINDOW = 5
        private const val WINDOW_FRAMES = 120   // 4s at 30fps
        private const val MIN_INIT_FRAMES = 45  // ~1.5s warmup after countdown ends
        private const val DOWN_FRACTION = 0.45f
        private const val UP_FRACTION = 0.20f
        private const val MIN_BAND = 0.06f
        private const val PERCENTILE_LOW = 0.15f
        private const val PERCENTILE_HIGH = 0.85f
        private const val MIN_UP_FRAMES = 9
        private const val MIN_DOWN_FRAMES = 11
        // Seeded band on warmup exit so the first rep counts before DOWN frames accumulate
        private const val INIT_BAND_SEED = 0.15f
    }

    fun start() {
        state = PushUpState.WARMING
        count = 0
        smoothBuf.clear()
        window.clear()
        lastSmoothed = 0f
        stateFrames = 0
        yLow = 0.5f
        yHigh = 0.5f
    }

    fun reset() {
        state = PushUpState.IDLE
        count = 0
        smoothBuf.clear()
        window.clear()
        lastSmoothed = 0f
        stateFrames = 0
        yLow = 0.5f
        yHigh = 0.5f
    }

    fun process(shoulderY: Float): Boolean {
        if (state == PushUpState.IDLE) return false

        val s = smooth(shoulderY).also { lastSmoothed = it }
        if (window.size >= WINDOW_FRAMES) window.poll()
        window.add(s)

        if (window.size < MIN_INIT_FRAMES) {
            stateFrames++
            return false
        }

        if (state == PushUpState.WARMING) {
            // Seed the initial band so the first rep is counted before DOWN frames
            // accumulate in the rolling window; without this, band stays near 0 until
            // the user has completed one full rep.
            val wSorted = window.sorted()
            val wn = wSorted.size
            yLow = wSorted[(wn * PERCENTILE_LOW).toInt()]
            yHigh = yLow + INIT_BAND_SEED
            state = PushUpState.UP
            stateFrames = 0
        }

        val sorted = window.sorted()
        val n = sorted.size
        val yl = sorted[(n * PERCENTILE_LOW).toInt()]
        // Maintain at least INIT_BAND_SEED band until real DOWN frames push yHigh higher
        val yh = maxOf(sorted[(n * PERCENTILE_HIGH).toInt()], yl + INIT_BAND_SEED)
        val band = yh - yl

        yLow = yl
        yHigh = yh

        if (band < MIN_BAND) {
            stateFrames++
            return false
        }

        val downThr = yl + DOWN_FRACTION * band
        val upThr   = yl + UP_FRACTION * band

        return when (state) {
            PushUpState.UP -> {
                stateFrames++
                if (s > downThr && stateFrames >= MIN_UP_FRAMES) {
                    state = PushUpState.DOWN
                    stateFrames = 0
                }
                false
            }
            PushUpState.DOWN -> {
                stateFrames++
                if (stateFrames >= MIN_DOWN_FRAMES && s < upThr) {
                    state = PushUpState.UP
                    count++
                    stateFrames = 0
                    true
                } else false
            }
            else -> false
        }
    }

    fun currentDisplacement(): Float = lastSmoothed - yLow

    fun downThresholdValue(): Float {
        val band = yHigh - yLow
        return yLow + DOWN_FRACTION * band
    }

    fun upThresholdValue(): Float {
        val band = yHigh - yLow
        return yLow + UP_FRACTION * band
    }

    fun warmupProgress(): Float =
        if (state == PushUpState.WARMING) window.size.toFloat() / MIN_INIT_FRAMES
        else 1f

    private fun smooth(y: Float): Float {
        if (smoothBuf.size >= SMOOTH_WINDOW) smoothBuf.poll()
        smoothBuf.add(y)
        return smoothBuf.average().toFloat()
    }
}
